plugins/inputs/snmp/snmp.go

```diff
@@ -0,0 +1,447 @@
+	return node.name, ""
+}
+
+func (s *Snmp) Gather(acc inputs.Accumulator) error {
+	// Create oid tree
+	if s.SnmptranslateFile != "" && len(initNode.subnodes) == 0 {
+		data, err := ioutil.ReadFile(s.SnmptranslateFile)
+		if err != nil {
+			log.Printf("Reading SNMPtranslate file error: %s", err)
+			return err
+		} else {
+			for _, line := range strings.Split(string(data), "\n") {
+				oidsRegEx := regexp.MustCompile(`([^\t]*)\t*([^\t]*)`)
+				oids := oidsRegEx.FindStringSubmatch(string(line))
+				if oids[2] != "" {
+					oid_name := oids[1]
+					oid := oids[2]
+					fillnode(initNode, oid_name, strings.Split(string(oid), "."))
+					NameToOid[oid_name] = oid
+				}
+			}
+		}
+	}
+	// Fetching data
+	for _, host := range s.Host {
+		// Set default args
+		if len(host.Address) == 0 {
+			host.Address = "127.0.0.1:161"
+		}
+		if host.Community == "" {
+			host.Community = "public"
+		}
+		if host.Timeout <= 0 {
+			host.Timeout = 2.0
+		}
+		if host.Retries <= 0 {
+			host.Retries = 2
+		}
+		// Prepare host
+		for _, oid_name := range host.Collect {
+			// Get GET oids
+			for _, oid := range s.Get {
+				if oid.Name == oid_name {
+					if val, ok := NameToOid[oid.Oid]; ok {
+						// TODO should we add the 0 instance ?
+						if oid.Instance != "" {
+							oid.rawOid = "." + val + "." + oid.Instance
+						} else {
+							oid.rawOid = "." + val + ".0"
+						}
+					} else {
+						oid.rawOid = oid.Oid
+					}
+					host.getOids = append(host.getOids, oid)
+				}
+			}
+			// Get GETBULK oids
+			for _, oid := range s.Bulk {
+				if oid.Name == oid_name {
+					if val, ok := NameToOid[oid.Oid]; ok {
+						oid.rawOid = "." + val
+					} else {
+						oid.rawOid = oid.Oid
+					}
+					host.bulkOids = append(host.bulkOids, oid)
+				}
+			}
+		}
+		// Launch Get requests
+		if err := host.SNMPGet(acc); err != nil {
+			return err
+		}
+		if err := host.SNMPBulk(acc); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (h *Host) SNMPGet(acc inputs.Accumulator) error {
+	// Get snmp client
+	snmpClient, err := h.GetSNMPClient()
+	if err != nil {
+		return err
+	}
+	// Prepare OIDs
+	oidsList := make(map[string]Data)
+	for _, oid := range h.getOids {
+		oidsList[oid.rawOid] = oid
+	}
+	oidsNameList := make([]string, 0, len(oidsList))
+	for _, oid := range oidsList {
+		oidsNameList = append(oidsNameList, oid.rawOid)
+	}
+
+	// gosnmp.MAX_OIDS == 60
+	// TODO use gosnmp.MAX_OIDS instead of hard coded value
+	max_oids := 60
+	// limit 60 (MAX_OIDS) oids by requests
+	for i := 0; i < len(oidsList); i = i + max_oids {
+		// Launch request
+		max_index := i + max_oids
+		if i+max_oids > len(oidsList) {
+			max_index = len(oidsList)
+		}
+		result, err3 := snmpClient.Get(oidsNameList[i:max_index]) // Get() accepts up to g.MAX_OIDS
+		if err3 != nil {
+			return err3
+		}
+		// Handle response
+		_, err = h.HandleResponse(oidsList, result, acc)
+		if err != nil {
+			return err
+		}
+	}
+	// Deconnection
+	defer snmpClient.Conn.Close()
```

Comment:
This defer should be moved up to line 264, directly after the client is made (but after the error is checked)